File: core0/us/sdk/wallet/java/us/wallet/trader/trades_index_t.java
//===-                           S C R I P T  T.V.
//===-                           https://script.tv
//===-
//===-            Copyright (C) 2017-2024 Script Network
//===-            Copyright (C) 2017-2024 manicberet@gmail.com
//===-
//===-                      GNU GENERAL PUBLIC LICENSE
//===-                       Version 3, 29 June 2007
//===-
//===-    This program is free software: you can redistribute it and/or modify
//===-    it under the terms of the GPLv3 License as published by the Free
//===-    Software Foundation.
//===-
//===-    This program is distributed in the hope that it will be useful,
//===-    but WITHOUT ANY WARRANTY; without even the implied warranty of
//===-    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//===-
//===-    You should have received a copy of the General Public License
//===-    along with this program, see LICENCE file.
//===-    see https://www.gnu.org/licenses
//===-
//===----------------------------------------------------------------------------
//===-
package us.wallet.trader;
import us.gov.io.blob_reader_t;                                                                // blob_reader_t
import us.gov.io.blob_writer_t;                                                                // blob_writer_t
import us.CFG;                                                                                 // CFG
import static us.gov.io.types.*;                                                               // *
import static us.ko.is_ko;                                                                     // is_ko
import us.ko;                                                                                  // ko
import static us.ko.ok;                                                                        // ok
import static us.stdint.*;                                                                        // ok
import us.pair;                                                                                // pair
import java.io.PrintStream;                                                                    // PrintStream
import us.gov.io.seriable;                                                                     // seriable
import static us.gov.io.types.blob_t.serid_t;                                                  // serid_t
import java.util.ArrayList;                                                                    // ArrayList
import static us.gov.crypto.ripemd160.hash_t;

@SuppressWarnings("serial")
public class trades_index_t extends ArrayList<trader_brief_t> implements us.gov.io.seriable {

    static void log(final String line) {                               //--strip
       CFG.log_wallet_trader("trades_index_t " + line);                //--strip
    }                                                                  //--strip

    @Override public serid_t serial_id() { return serid_t.no_header; }

    @Override public int blob_size() {
        int sz = blob_writer_t.sizet_size(size());
        for (trader_brief_t o: this) {
            sz += o.blob_size();
        }
        return sz;
    }

    @Override public void to_blob(blob_writer_t writer) {
        log("to_blob " + size() + " entries"); //--strip
        writer.write_sizet(size());
        for (trader_brief_t o: this) {
            o.to_blob(writer);
        }
    }

    @Override public ko from_blob(blob_reader_t reader) {
        clear();
        log("from_blob"); //--strip
        uint64_t sz = new uint64_t();
        {
            ko r = reader.read_sizet(sz);
            if (is_ko(r)) return r;
        }
        log("loading " + sz.value + " items from blob"); //--strip
        try {
            for (long i = 0; i < sz.value; ++i) {
                log("loading item " + i); //--strip
                trader_brief_t trader_brief = new trader_brief_t();
                {
                    ko r = trader_brief.from_blob(reader);
                    if (is_ko(r)) {
                        log(r.msg); //--strip
                        return r;
                    }
                }
                log("adding entry"); //--strip
                add(trader_brief);
            }
        }
        catch (Exception e) {
            return new ko("KO 82282 Invalid read.");
        }
        return ok;
    }

    
}

File: core0/us/wallet/trader/trades_index_t.cpp
//===-                           S C R I P T  T.V.
//===-                           https://script.tv
//===-
//===-            Copyright (C) 2017-2024 Script Network
//===-            Copyright (C) 2017-2024 manicberet@gmail.com
//===-
//===-                      GNU GENERAL PUBLIC LICENSE
//===-                       Version 3, 29 June 2007
//===-
//===-    This program is free software: you can redistribute it and/or modify
//===-    it under the terms of the GPLv3 License as published by the Free
//===-    Software Foundation.
//===-
//===-    This program is distributed in the hope that it will be useful,
//===-    but WITHOUT ANY WARRANTY; without even the implied warranty of
//===-    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//===-
//===-    You should have received a copy of the General Public License
//===-    along with this program, see LICENCE file.
//===-    see https://www.gnu.org/licenses
//===-
//===----------------------------------------------------------------------------
//===-
#include "trades_index_t.h"

#define loglevel "wallet/trader"
#define logclass "trades_index_t"
#include <us/gov/logs.inc>

using namespace us::wallet::trader;
using d = us::wallet::trader::trader_brief_t;

size_t d::blob_size() const {
    auto sz = blob_writer_t::blob_size(tid)
    + blob_writer_t::blob_size(ts_creation)
    + blob_writer_t::blob_size(ts_activity)
    + blob_writer_t::blob_size(subhome)
    + blob_writer_t::blob_size(qr);
    log("blob_size", sz);
    return sz;
}

void d::to_blob(blob_writer_t& writer) const {
    log("to_blob", "cur", (uint64_t)(writer.cur - writer.blob.data()));
    writer.write(tid);
    writer.write(ts_creation);
    writer.write(ts_activity);
    writer.write(subhome);
    writer.write(qr);
}

ko d::from_blob(blob_reader_t& reader) {
    log("from_blob", "cur", (uint64_t)(reader.cur - reader.blob.data()));
    {
        auto r = reader.read(tid);
        if (is_ko(r)) {
            return r;
        }
    }
    {
        auto r = reader.read(ts_creation);
        if (is_ko(r)) {
            return r;
        }
    }
    {
        auto r = reader.read(ts_activity);
        if (is_ko(r)) {
            return r;
        }
    }
    {
        auto r = reader.read(subhome);
        if (is_ko(r)) {
            return r;
        }
    }
    {
        auto r = reader.read(qr);
        if (is_ko(r)) {
            return r;
        }
    }
    return ok;
}

/*
void c::dump(const string& pfx, ostream& os) const {
    lock_guard<mutex> lock(mx);
    for (auto& i: *this) {
        os << pfx << i.first << ' ' << i.second << '\n';
    }
}
*/

